### PR TITLE
Stop producing and handling `ADDR` nodes

### PR DESCRIPTION
**Pull request by SingleAccretion**
With this change, the `GT_ADDR` oper (for a short time before its eventual deletion) becomes "synthetic": no actual nodes of it will be created.

We also start using local address nodes in more places and delete dead code.

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=108382&view=ms.vss-build-web.run-extensions-tab): minor GC info size reductions (`LCL_VAR_ADDR`s are typed `I_IMPL`); some substitution diffs (due to fewer `GLOB_REF`s).

There are nice TP improvements: `0.3%` on 64 bit platforms, `1%` on 32 bit platforms, of which the latter comes mostly from the `OperIsIndir` fix.

### CODE REVIEW

File: src/coreclr/jit/gentree.cpp [hunk 1]
```diff
@@ -16125,18 +16115,6 @@ void Compiler::gtExtractSideEffList(GenTree*     expr,
                     return Compiler::WALK_SKIP_SUBTREES;
                 }
 
-                if ((m_flags & GTF_EXCEPT) != 0)
-                {
-                    // Special case - GT_ADDR of GT_IND nodes of TYP_STRUCT have to be kept together.
```
Comment (SingleAccretion) [inline src/coreclr/jit/gentree.cpp:16130]: Leftover from `GT_INDEX` that I forgot to delete.

File: src/coreclr/jit/gentree.h [hunk 2]
```diff
@@ -1564,7 +1564,9 @@ struct GenTree
     // OperIsIndir() returns true also for indirection nodes such as GT_BLK, etc. as well as GT_NULLCHECK.
     static bool OperIsIndir(genTreeOps gtOper)
     {
-        return gtOper == GT_IND || gtOper == GT_STOREIND || gtOper == GT_NULLCHECK || OperIsBlk(gtOper);
```
Comment (drieseng) [inline src/coreclr/jit/gentree.h:1567]: Are such issues ever reported to the MSVC team?
Perhaps consider adding a comment with a link to the reported issue.
Feel free to ignore these remarks, just passing by and watching your work in awe 😀

**Review by BruceForstall — APPROVED** [review]
Changes LGTM. Thanks for your continuing valuable contributions!

I kicked off stress runs.

**Comment by BruceForstall** [comment]
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=108382&view=ms.vss-build-web.run-extensions-tab)

**Comment by BruceForstall** [comment]
libraries-jitstress had a few failures: one known, one looked like infra, Linux/arm tailcallstress had a "new" crash in System.Memory.Tests. Set to rerun to see if it repros.

**Comment by BruceForstall** [comment]
Didn't repro; all failures are "known".

**Pull request merged**